db/records/operations/group.py

```diff
@@ -1,123 +1,208 @@
-from sqlalchemy import select, Column, func
+from dataclasses import dataclass, field
+from enum import Enum
+import json
+import logging
+from typing import List, Tuple
+from sqlalchemy import select, Column, func, and_, case, literal
 
-from db.records.exceptions import BadGroupFormat, GroupFieldNotFound
-from db.records.operations.select import get_query, apply_filters
+from db.records import exceptions as rec_exc
 from db.records.utils import create_col_objects
-from db.utils import execute_query
 
+logger = logging.getLogger(__name__)
 
-def append_distinct_tuples_to_filter(distinct_tuples):
-    filters = []
-    for col, value in distinct_tuples:
-        filters.append({
-            "field": col,
-            "op": "==",
-            "value": value,
-        })
-    return filters
+MATHESAR_GROUP_METADATA = '__mathesar_group_metadata'
 
 
-def get_distinct_tuple_values(
-        column_list, engine, table=None, limit=None, offset=None, output_table=None
-):
+class GroupMode(Enum):
+    DISTINCT = 'distinct'
+    PERCENTILE = 'percentile'
+
+
+class GroupMetadataField(Enum):
+    COUNT = 'count'
+    GROUP_ID = 'group_id'
+    FIRST_VALUE = 'first_value'
+    LAST_VALUE = 'last_value'
+
+
+@dataclass(frozen=True, eq=True)
+class GroupBy:
+    column_list: List
+    group_mode: str = GroupMode.DISTINCT.value
+    num_groups: int = 12
+
+    def validate(self):
+        if type(self.column_list) not in (tuple, list):
+            raise rec_exc.BadGroupFormat("column_list must be list or tuple.")
+        if self.group_mode not in {mode.value for mode in GroupMode}:
+            raise rec_exc.InvalidGroupType(
+                f'Group_mode "{self.group_mode}" is invalid.'
+            )
+        if (
+                self.group_mode == GroupMode.PERCENTILE.value
+                and not type(self.num_groups) == int
+        ):
+            raise rec_exc.BadGroupFormat(
+                'percentile group_mode requires integer num_groups'
+            )
+
+        for col in self.column_list:
+            if type(col) not in (str, Column):
+                raise rec_exc.BadGroupFormat(
+                    f"Group column {col} must be a string or Column."
+                )
```

Comment:
I think you're right that we can require string args, and avoid allowing `Column`.  I think the option was from some logic (that appears to be missing now) that would allow for columns from different tables. 

I don't understand your final comment.  The `GroupBy.group_mode` is checked to be a string by the `validate` function (since each value of the `GroupMode` Enum is a string).

Keep in mind that I'm not actually convinced that we should actually _use_ the type annotations for these dataclasses.  I find it a bit awkward that they're semi-required by the dataclasses themselves, but we're not using them anywhere else in the codebase.

Reply:
I confused `column_list` and `group_mode` fields. I meant to say that `column_list` should be `List[Union[str, Column]]` according to the `validate` method. I personally like type annotations. I find them useful, because of readability, but also because of the Language Server Protocol server I'm using: it gives me in-editor warnings and errors if I make a mistake. I like that they're optional too. I don't always annotate methods and functions, but I always try to annotate data structures.